### ISSUE TITLE
Typechecker: Matchings that yield optionals cannot infer the type of `None`

### ISSUE THREAD

**Issue by greytdepression** [SerenityOS/jakt#669]
The following code fails to compile, because the typechecker does not know how to handle `None` here.
```jakt
function foo() -> bool? => match 0 {
    1 => false
    else => None
}

function main() {
    let a = foo()
}
```
The error (in which the types are flipped, but that's for another issue :^) ) is
```
Error: Type mismatch: expected ‘unknown’, but got ‘bool’
 10 |     false => false 
 11 |     else => None 
                  ^- Type mismatch: expected ‘unknown’, but got ‘bool’
 12 | } 
-----
```

**Comment by steven-mathew** [comment]
Hey there! I'd like to take a look into this. I'll reach out on discord with any questions I have.

**Comment by lanmonster** [comment]
fixed as of 1ea81c6. filed #1139 as a follow up for type inferrence.
